modules/ocf_ldap/files/ldap-lint

```diff
@@ -0,0 +1,190 @@
+        dns_ip = lookup_dns(cn + '.ocf.berkeley.edu')
+
+        if not dns_ip:
+            complain(cn, 'has no A record in DNS')
+        else:
+            if ip in seen_ips:
+                complain(cn, 'has same IP address as ' + seen_ips[ip])
+            else:
+                seen_ips[ip] = cn
+
+            if dns_ip != ip:
+                complain(cn, 'ldap ip {} doesn\'t match dns ip {}'.format(ip, dns_ip))
+
+        ptr = lookup_dns(dns.reversename.from_address(ip), rtype='PTR')
+
+        if ptr:
+            if ptr.lower() != cn + '.ocf.berkeley.edu.':
+                complain(cn, 'bad reverse DNS for {}: {}'.format(ip, ptr))
+        else:
+            complain(cn, 'missing reverse DNS for {}'.format(ip))
+
+        # TODO: remove these from schema
+        for puppet_var in attrs.get('puppetVar', []):
+            complain(cn, 'has puppetVar: {}'.format(puppet_var))
+
+        for puppet_class in attrs.get('puppetClass', []):
+            complain(cn, 'has puppetClass: {}'.format(puppet_class))
+
+
+def get_kadmin_users(filter_query):
+    """Using kadmin -l as root means that we don't have to get a kerberos
+    principal for this script and it can always check the local kerberos, which
+    is nice"""
+    process = subprocess.run(
+        ('kadmin', '-l', 'list', filter_query),
+        stdout=subprocess.PIPE,
+    )
+    principals = process.stdout.decode('utf-8').split('\n')
+
+    # Only return usernames, not the full principal name, since the suffix
+    # isn't useful
+    return set([principal.split('/')[0] for principal in principals])
+
+
+def get_users_in_group(c, group):
+    """This is useful over the (much simpler) list_group from ocflib since it
+    uses the local LDAP instead of the cached user groups, so it validates the
+    actual source of truth and can also work on dev-ldap"""
+    c.search(
+        OCF_LDAP_GROUP,
+        '(cn={})'.format(group),
+        attributes=['memberUid'],
+    )
+    return set(c.response[0]['attributes']['memberUid'])
+
+
+def check_users(c, complain):
+    ocfstaff = get_users_in_group(c, 'ocfstaff')
+    ocfroot = get_users_in_group(c, 'ocfroot')
+
+    admin_users = get_kadmin_users('*/admin')
+    root_users = get_kadmin_users('*/root')
+
+    for user in (admin_users - root_users):
+        if user not in ('create', 'kadmin'):
+            complain(user + '/admin', 'no corresponding {}/root principal'.format(user))
+
+    for user in (root_users - admin_users):
+        complain(user + '/root', 'no corresponding {}/admin principal'.format(user))
+
+    for user in ocfroot:
+        if user not in ocfstaff:
+            complain(user, 'in ocfroot but not ocfstaff')
+
+        if user not in root_users:
+            complain(user, 'in ocfroot but has no root principal')
+
+        if user not in admin_users:
+            complain(user, 'in ocfroot but has no admin principal')
+
+
+def main():
+    retval = 0
+
+    def complain(cn, error):
```

Comment:
can't we just... define this at the top level to avoid having to awkwardly pass it into `check_hosts` and `check_users`?

Reply:
The issue with that is that it also sets `retval` along the way. We could just make that a global and that'd work though, so that's probably just worth doing.

Reply:
From trying this out I'm not sure it's really much cleaner. Yes the function isn't nested, but you still need the `nonlocal retval`, it's just essentially not passing it into two functions and moving it further up in the file...

Reply:
That would still be my preference but I see your point.